src/lib/Simplify.hs

```diff
@@ -39,6 +39,62 @@ import LabeledItems
 import Types.Primitives
 import Types.Core
 
+-- === Simplification ===
+
+-- The purpose of simplification is that we want first-class functions
+-- in the surface language, but we don't want to deal with them when
+-- generating code.  To that end, simplification inlines almost all
+-- functions everywhere they appear.
+
+-- In particular, simplification also carries out AD by discharging
+-- the `Linearize` and `Transpose` constructors of `PrimHof`.
+
+-- The exception is functions explicitly tagged @noinline by the
+-- programmer.  Those, however, are second-class: they are all
+-- toplevel, and get specialized until they are first order.
+
+-- Currently, simplification also discharges `CatchException` by
+-- elaborating the expression into a Maybe-style monad.  Note: the
+-- plan is for `CatchException` to become a user-defined effect, and
+-- for simplification to discharge all of them.
+
+-- Simplification also discharges bulk record and variant operations
+-- by converting them into individual projections and constructors.
+-- TODO(reviewer): Can it always discharge all of them?
+
+-- Simplification also opportunistically does peep-hole optimizations:
+-- some constant folding, case-of-known-constructor, projecting known
+-- elements from products, etc; but is not guaranteed to find all such
+-- opportunities.
+
+-- Simplification happens after dictionary synthesis (`synthTopBlock`)
+-- and before lowering.  Its input invariants are therefore
+-- - PrimCon has no DictHole constructors
+-- - PrimOp has no
+--   - Low-level memory operations
+--   - Destination operations
+--   - Vector operations
```

Comment:
I think vector operations are inessential here (simplification would handle them as written), while low-level memory and destination ops all belong to the "Dex abstract machine" umbrella IMO